### PR TITLE
Decouple parmed from the foyer atomtyper

### PR DESCRIPTION
**Pull request by umesh-timalsina**
This issue has been discussed at length in #386. Specifically this PR is intended to address https://github.com/mosdef-hub/foyer/issues/386#issuecomment-794684790 and with #382 underway. I think we can safely commit to removing third party packages for atomtyping engine from foyer.

Checklist:
- [x] Remove parmed reference from `atom_typer.py`
- [x] Remove atom references in `smarts_graph.py`
- [x] Build a separate utility that subclasses nx.Graph, which has to have certain attributes needed for atom_typing? Discussion needed.

Based on the third point a utility class could look like the following:

```py
import networkx as nx
from parmed import Structure


class AtomData:
    def __init__(self, index, name, atomic_number, element, bond_partners, **kwargs):
        self.index = index
        self.name = name
        self.atomic_number = atomic_number
        self.element = element
        self.bond_partners = bond_partners
        for key, value in kwargs.items():
            setattr(self, key, value)


class TopologyGraph(nx.Graph):
    def __init__(self, *args, **kwargs):
        super(TopologyGraph, self).__init__(*args, **kwargs)

    def add_atom(self, index, name, atomic_number, element, bond_partners, **kwargs):
        atom_data = AtomData(
            index, name, atomic_number, element, bond_partners, **kwargs
        )
        self.add_node(index, atom_data=atom_data)

    def add_bond(self, atom_1_index, atom_2_index):
        self.add_edge(atom_1_index, atom_2_index)

    def atoms(self, data=False):
        if data:
            for idx, data in self.nodes(data=data):
                yield idx, data["atom_data"]
        else:
            for idx in self.nodes(data=data):
                yield idx

    @classmethod
    def from_parmed(cls, structure: Structure) -> nx.Graph:
        """Return a networkx graph with relevant attributes from a parmed Structure"""
        topology_graph = cls()
        for atom in structure.atoms:
            topology_graph.add_atom(
                name=atom.name,
                index=atom.idx,
                atomic_number=atom.atomic_number,
                element=atom.element,
                bond_partners=[bonded_atom.idx for bonded_atom in atom.bond_partners],
            )

        for bond in structure.bonds:
            topology_graph.add_bond(bond.atom1.idx, bond.atom2.idx)
        return topology_graph
```

### CODE REVIEW

File: foyer/atomtyper.py [hunk 1]
```diff
@@ -11,42 +11,55 @@ def find_atomtypes(structure, forcefield, max_iter=10):
 
     Parameters
     ----------
-    topology : simtk.openmm.app.Topology
+    topology : parmed.Structure of nx.Graph
```
Comment (umesh-timalsina) [inline foyer/atomtyper.py:14]:
```suggestion
    topology : parmed.Structure or nx.Graph
```

**Comment by umesh-timalsina** [comment]
This PR is now complete for initial round of reviews.

**Review by daico007 — APPROVED** [review]
LGTM!

File: foyer/topology_graph.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    **kwargs
+        Any extra information for this atom
+
+    ToDo: After 3.6 support removed, convert this to a dataclass
+    """
+
+    def __init__(self, index, name, atomic_number, element, bond_partners, **kwargs):
+        self.index = index
+        self.name = name
+        self.atomic_number = atomic_number
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply foyer/topology_graph.py:29]: For `non_element` atom types, this attribute is not used.

File: foyer/topology_graph.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        **kwargs
+            Any extra information for this atom
+
+        See Also
+        --------
+        foyer.topology_graph.AtomData
+            The class used to store atom data.
+        """
+        atom_data = AtomData(
+            index, name, atomic_number, element, bond_partners, **kwargs
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply foyer/topology_graph.py:71]: This was done to make minimal changes to the existing API. I could change it grab this information from the node itself.

File: foyer/topology_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply foyer/topology_graph.py:29]: But I see your point. Will change this.

File: foyer/topology_graph.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        TopologyGraph
+            The equivelent TopologyGraph of the parmed Structure `structure`
+        """
+        topology_graph = cls()
+        for atom in structure.atoms:
+            topology_graph.add_atom(
+                name=atom.name,
+                index=atom.idx,
+                atomic_number=atom.atomic_number,
+                element=atom.element,
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply foyer/topology_graph.py:115]: Agreed.

File: foyer/atomtyper.py [hunk 5]
```diff
@@ -11,42 +13,54 @@ def find_atomtypes(structure, forcefield, max_iter=10):
...
 
     system_elements = set()
-    for a in structure.atoms:
+    for _, atom_data in topology_graph.atoms(data=True):
         # First add non-element types, which are strings, then elements
-        if a.name.startswith('_'):
-            if a.name in forcefield.non_element_types:
-                system_elements.add(a.name)
+        name = atom_data.name
+        atomic_number = atom_data.atomic_number
```
Comment (rsdefever) [inline foyer/atomtyper.py]: This line is why I was asking about how non-element atom types handle `atomic_number`.

File: foyer/topology_graph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply foyer/topology_graph.py]: This information is now build on the fly.

**Pull request merged**